frontend/src/search.js

```diff
@@ -83,6 +83,34 @@ search.addWidgets([
         document.getElementById('under-progress').classList.remove('hidden');
     },
   },
+  {
+    message: '',
+    init() {
+      document.getElementById('retry-search').addEventListener('click', () => {
+        document.getElementById('loading').classList.remove('hidden');
+        document.getElementById('stats').classList.add('hidden');
+        document.getElementById('error').classList.add('hidden');
+        this.message = '';
+        setTimeout(() => {
+          search.helper.search();
+        }, 500);
+      });
+      search.on('error', ({ error }) => {
+        this.message = error.message;
+        document.getElementById('error-message').innerHTML = error.message;
+      });
+    },
+    render({ searchMetadata = {} }) {
+      const { isSearchStalled } = searchMetadata;
+      document
+        .getElementById('stats')
+        .classList.toggle('hidden', this.message || isSearchStalled);
```

Comment:
Let's try top avoid manipulating `stats` widget from here and from the `loading` widget as well. We can do this by customising the `stats` widget itself: https://www.algolia.com/doc/api-reference/widgets/stats/js/#connector. It can react on the error/stalled states and hide itself.